### PR TITLE
feat: Timecode & scheduler

### PR DESCRIPTION
**Pull request by Julusian**
#871 #872 #1004

This adds support for timecode input and output via decklinks, and for being able to schedule commands to be executed at various timecodes. It also adds a batch command syntax to amcp.

To be able to use a decklink as the source, it must exist on the stage. To allow this from startup, the xml config can now have producers defined.

Further docs: https://docs.google.com/document/d/1xkksxIdNHMO95RFF8wJ3vYRbXCHwKnjkMwVqwAyjv3A/edit?usp=sharing

### CODE REVIEW

**Comment by ronag** [comment]
Can you put the AMCP updates into a separate PR?

File: src/modules/decklink/producer/decklink_producer.cpp [hunk 1]
```diff
@@ -509,6 +509,12 @@ class decklink_producer : public IDeckLinkInputCallback
                 }
             }
 
+            if (new_timecode != core::frame_timecode::empty())
+            {
+                // Account for latency introduced by ffmpeg filter
+                new_timecode = new_timecode - (format_desc_.field_count > 1 ? 3 : 1);
```
Comment (ronag) [inline src/modules/decklink/producer/decklink_producer.cpp]: This is not going to be entirely accurate...

Comment (ronag) [reply src/modules/decklink/producer/decklink_producer.cpp]: Hm, no that's a bad idea, but I still think you can measure the filter latency using pts

Comment (ronag) [reply src/modules/decklink/producer/decklink_producer.cpp]: hmm... maybe this is good enough in practice

Comment (ronag) [reply src/modules/decklink/producer/decklink_producer.cpp]: leave it

**Comment by ronag** [comment]
@Julusian I have it on my list to go through this... can you please resolve the conflicts?

**Comment by didikunz** [comment]
Hi, is there a build with that included? It looks to me, that it's still not merged into the main branch, right?

**Comment by ronag** [comment]
@didikunz: Not merged. There are lot's of conflicts.

**Comment by didikunz** [comment]
I would encourage @Julusian or who ever is responsible to resolve this, as this is a power full feature. Is it in the NHK builds?

**Comment by MauriceVeraart** [comment]
Yes

**Comment by swordh** [comment]
Hi Guys,
Love the work, keep it up :)  waiting for this feature to be merged, only reason that I use the NRK version and Windows. 
Are there any plans on getting this merged soon?

**Comment by poldon** [comment]
Is there any plan to merge this still?  This is a very good feature, and one I think a lot of people will benefit from.

**Review by GeorgNation — CHANGES_REQUESTED** [review]
I think it should be in release. The only thing missing is the ability to switch the time zone, timecode with a drop frame, specifying milliseconds and the ability to hold an event. This would make the tool ideal.

**Comment by Julusian** [comment]
I think it is time to close this PR, discussion can continue here or in the issue #871 

The client who originally funded this work no longer has an interest in it, and some more work is needed to flesh out timecode support to make it more usable.
I am not entirely sure on the resulting quality of this change as it was never used in production (other than with the default config of following the system clock). For this to be actually useful to anyone, I expect that support will need to be added to the ffmpeg_producer, and possibly more producers and consumers.

I've had a quick attempt at rebasing this onto current builds, as there were pretty major merge conflicts due to various portions of this having been picked and merged already. If anyone wants to continue working on this you can start with https://github.com/CasparCG/server/commit/056c9469da08b5c958ccf5fc12713e62a94490fa

I have no interest in the scheduler portion of this PR, we found it to be hard to work with and to understand the state of the system when cancelling commands or amending the scheduled commands.

**Pull request closed without merging**